Explain malformed tracks and what to do with them (#938)
moq-wg/moq-transport: draft-ietf-moq-transport.md

UTF-8. Any specification that does needs to specify the canonicalization into
the bytes in the Track Namespace or Track Name such that exact comparison works.

## Malformed Tracks

A track is considered malformed if any of the following conditions are detected
by a receiver:

1. An Object is received on a Subgroup whose Object ID is not strictly larger
   than the previous Object received on the same Subgroup.
2. An Object is received in a FETCH response with the same Group as the
   previous Object, but whose Object ID is not strictly larger than the previous
   object.

> 3. An Object is received in an Asending FETCH response whose Group ID is smaller

Reviewer: ```suggestion
3. An Object is received in an Ascending FETCH response whose Group ID is smaller
```

   than the previous Object in the response.
4. An Object is received in a Descending FETCH response whose Group ID is larger
   than the previous Object in the resopnse.
5. A Subgroup or FETCH response is terminated with a FIN in the middle of an
   Object
6. An Object is received whose Object ID is larger than the confirmed largest

>    Object in the subgroup.

Reviewer 2: I don't think i fully understand "confirmed largest object" and also scoped to a "subgroup".
By confirmed , are we saying as seen by a Subscriber in OK message and why is it an error if that subscriber receives larger object ?

Also why is this scoped to a subgroup rather than to a track ?

Author: Yeah I need to explain confirmed better.  The "confirmed largest" thing means one of three things:

1) Receiver has processed a FETCH that showed there was nothing larger in that subgroup/group/track
2) Receiver has processed an "end of..." Object Status for the Track or Group
3) Receiver has processed a subgroup stream with a FIN

Author: @suhasHere I changed the language to call it "final Object" in the (Subgroup/Group/Track) and tried to define it better, and also added multiple final objects in a subgroup as a malformed condition.

7. An Object is received in a Group whose Object ID is larger than the
   confirmed largest Object in the Group.

> 8. An Object is received on a Track whose Object ID is larger than the confirmed
>    largest Object in the Track.
> 8. The same Object is received more than once with with different Payload or
>    other immutable properties.

Reviewer: ```suggestion
8. An Object is received on a Track whose Group and Object ID are larger than the\
   confirmed largest Object in the Track.
9. The same Object is received more than once with with different Payload or
   other immutable properties.
```

UTF-8. Any specification that does needs to specify the canonicalization into
the bytes in the Track Namespace or Track Name such that exact comparison works.

## Malformed Tracks

A track is considered malformed if any of the following conditions are detected
by a receiver:

> 1. An Object is received on a Subgroup whose Object ID is not strictly larger

Reviewer 3: This does not apply if using Datagram delivery.

Reviewer 3: So for clarity should this say "stream" somewhere in this bullet as it appears to be only be in the moment of receiving the subgroup on a stream?

   than the previous Object received on the same Subgroup.
2. An Object is received in a FETCH response with the same Group as the
   previous Object, but whose Object ID is not strictly larger than the previous
   object.
3. An Object is received in an Ascending FETCH response whose Group ID is smaller
   than the previous Object in the response.
4. An Object is received in a Descending FETCH response whose Group ID is larger
   than the previous Object in the resopnse.
5. A Subgroup or FETCH response is terminated with a FIN in the middle of an
   Object
6. An Object is received whose Object ID is larger than the final Object in the
   Subgroup.  The final Object in a Subgroup is the last Object received on a
   Subgroup stream before a FIN.

> 7. A Subgroup is received with two or more different final Objects.

Author: For example, publisher opens stream for sg=1, sends object=0, FIN, then opens a second stream for sg=1, sends object=1, FIN.

UTF-8. Any specification that does needs to specify the canonicalization into
the bytes in the Track Namespace or Track Name such that exact comparison works.

## Malformed Tracks

> A track is considered malformed if any of the following conditions are detected
> by a receiver:

Author: ```suggestion
There are multiple ways a publisher can transmit a Track that does not conform
to MoQT constraints. Such a Track is considered malformed.  Some example
conditions that constitute a malformed track when detected by a receiver
include:
```

> 1. An Object is received on a Subgroup whose Object ID is not strictly larger
>    than the previous Object received on the same Subgroup.

Author: ```suggestion
1. An Object is received on a Subgroup stream whose Object ID is not strictly
   larger than the previous Object received on the same Subgroup.
```

   than the previous Object in the resopnse.
5. A Subgroup or FETCH response is terminated with a FIN in the middle of an
   Object
6. An Object is received whose Object ID is larger than the final Object in the
   Subgroup.  The final Object in a Subgroup is the last Object received on a
   Subgroup stream before a FIN.
7. A Subgroup is received with two or more different final Objects.
8. An Object is received in a Group whose Object ID is larger than the final
   Object in the Group.  The final Object in a Group is the Object with Status
   END_OF_GROUP or the last Object sent in a FETCH that requested the entire
   Group.
9. An Object is received on a Track whose Group and Object ID are larger than the
   final Object in the Track.  The final Object in a Track is the Object with
   Status END_OF_TRACK or the last Object sent in a FETCH whose response indicated
   End of Track.
10. The same Object is received more than once with with different Payload or
    other immutable properties.
11. An Object is received with a different Forwarding Preference than previously
    observed from the same Track.

> When a subscriber detects a Malformed Track, it MUST UNSUBSCRIBE from the

Author: ```suggestion
The above list of conditions is not considered exhaustive.

When a subscriber detects a Malformed Track, it MUST UNSUBSCRIBE from the
```

   previous Object, but whose Object ID is not strictly larger than the previous
   object.
3. An Object is received in an Ascending FETCH response whose Group ID is smaller
   than the previous Object in the response.
4. An Object is received in a Descending FETCH response whose Group ID is larger
   than the previous Object in the resopnse.
5. A Subgroup or FETCH response is terminated with a FIN in the middle of an
   Object
6. An Object is received whose Object ID is larger than the final Object in the
   Subgroup.  The final Object in a Subgroup is the last Object received on a
   Subgroup stream before a FIN.
7. A Subgroup is received with two or more different final Objects.
8. An Object is received in a Group whose Object ID is larger than the final
   Object in the Group.  The final Object in a Group is the Object with Status
   END_OF_GROUP or the last Object sent in a FETCH that requested the entire
   Group.
9. An Object is received on a Track whose Group and Object ID are larger than the
   final Object in the Track.  The final Object in a Track is the Object with
   Status END_OF_TRACK or the last Object sent in a FETCH whose response indicated
   End of Track.

> 10. The same Object is received more than once with with different Payload or

Reviewer 4: ```suggestion
10. The same Object is received more than once with different Payload or
```

* Too Far Behind - The publisher's queue of objects to be sent to the given
  subscriber exceeds its implementation defined limit.

> * Malformed Track - The publisher detected the track was malformed (see

Reviewer 2: This can be sent mainly by relay publisher after its subscription shows the upstream publisher had a bad track. I wonder if we need to make it explicit and say "Relay Publisher "

> * Malformed Track - The publisher detected the track was malformed (see
>   {{malformed-tracks}}).

Author: ```suggestion
* Malformed Track - A relay publisher detected the track was malformed (see
  {{malformed-tracks}}).
```
